feat: add custom toolbar and shopping cart slot

rkristelijn opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

![image](https://github.com/user-attachments/assets/defb3950-4d3a-4dfd-b028-293dbd1c0781)

Adds ability to ship your own toolbar, this gives more flexibility, therefore: you can ship your own things, this may as well be the tools Toolpad already provides, such as the account, toolbar and any custom stuff, like a cart, or a search bar, in any given order.

Please let me know if this is in line with the roadmap, or how I can polish my merge request to approval.

- [x] I've read and followed the [contributing guide](https://github.com/mui/toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [ ] I've updated the relevant documentation for any new or updated feature.
- [x] I've linked relevant GitHub issue with "Closes #4662".
- [x] I've added a visual demonstration in the form of a screenshot or video.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -58,11 +60,20 @@ export interface DashboardLayoutSlots {
    * @default Account
    */
   toolbarAccount?: React.JSXElementConstructor<AccountProps>;
+  /**
```
Instead of creating any new slots, could we just move the `<Account />` component to be included by default on the right of the `toolbarActions` slot, and get rid of the `toolbarAccount` slot?

Then anyone could customize these actions to have whatever they wanted in any order.
Some slight updates might also be needed to the documentation around this slot.

rkristelijn replied:
I'm unsure to simply 'drop' the toolbarAccount slot, shouldn't we deprecate it first?

apedroferreira replied:
Actually yeah, that might be better, we can just deprecate it!

rkristelijn replied:
@apedroferreira I just did that and added the documentation. Thanks for the help!

<img width="1415" alt="image" src="https://github.com/user-attachments/assets/03933686-1fd7-4ef0-a95f-2a9cdf71e7e9" />

apedroferreira commented on playground/nextjs-pages/.env.example:
```diff
@@ -0,0 +1,7 @@
+# Force development mode
```
No need to change anything in the playgrounds or examples for now unless it's really necessary.
I'd keep them as they are!

rkristelijn replied:
gotcha

apedroferreira reviewed (changes requested):
Thanks a lot for your contribution!
Added some suggestions that will probably help simplify everything and align with what we thinking would be the best way to do this.

rkristelijn commented:
@apedroferreira I lifted the ability to override to the full toolbar and introduced left, mid and right props too. I'm unsure if i should extend:

```tsx
export interface DashboardLayoutSlotProps {
  appTitle?: AppTitleProps;
  toolbarActions?: {};
  toolbarAccount?: AccountProps;
  sidebarFooter?: SidebarFooterProps;
  toolbar?: {}; // needed?
}
```

And I will dive into updating the docs once you approved of this direction/change. Thanks for your feedback!

rkristelijn commented:
I think it is good enough now;

```tsx
function Middle({ menuIcon }: ToolbarProps) {
  return (
    <Stack direction="row" spacing={2} alignItems="center" justifyContent="space-between">
      {menuIcon}
      <ToolbarActions />
    </Stack>
  );
}

function Right() {
  return (
    <Stack direction="row" spacing={2} alignItems="center">
      <Account />

      <Link href="/orders">
        <Button color="primary" aria-label="Cart">
          <ShoppingCart />
        </Button>
      </Link>
    </Stack>
  );
}

function Left() {
  return (
    <Stack direction="row" spacing={2} alignItems="center">
      <AppTitle />
    </Stack>
  );
}

export default function DashboardPagesLayout(props: { children: React.ReactNode }) {
  const pathname = usePathname();
  const params = useParams();
  const [orderId] = params.segments ?? [];

  const title = React.useMemo(() => {
    if (pathname === '/orders/new') {
      return 'New Order';
    }
    if (orderId && pathname.includes('/edit')) {
      return `Order ${orderId} - Edit`;
    }
    if (orderId) {
      return `Order ${orderId}`;
    }
    return undefined;
  }, [orderId, pathname]);

  return (
    <DashboardLayout
      slots={{
        sidebarFooter: SidebarFooterAccount,
        // toolbarAccount: () => null,
        // toolbar: Middle,
        // toolbarLeft: Middle,
        toolbarLeft: () => <div />,
        toolbarRight: () => <div />,
        toolbarMiddle: Middle,
        // toolbarRight: Right,
        // toolbarLeft: Left,
        // toolbarLeft: () => null,
        // toolbarRight: () => null,
      }}
    >
      <PageContainer title={title}>{props.children}</PageContainer>
    </DashboardLayout>
  );
}
```
renders as:
<img width="736" alt="image" src="https://github.com/user-attachments/assets/170fb261-161f-4ec0-b7aa-b113c181af91" />

rkristelijn commented:
@apedroferreira Is it good to go now? I don't know how to apply a label

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -341,52 +386,59 @@ function DashboardLayout(props: DashboardLayoutProps) {
-            <Stack direction="row" alignItems="center" spacing={1} sx={{ marginLeft: 'auto' }}>
-              <ToolbarActionsSlot {...slotProps?.toolbarActions} />
-              <ToolbarAccountSlot {...slotProps?.toolbarAccount} />
+          {slots?.toolbar ? (
```
I would not add a new slot for now, as we still haven't decided how customizations options for the header as a whole will evolve.

rkristelijn replied:
1. The appBar slot is designed to give users full control over the entire top bar
2. When using the appBar slot, users can:
- Completely customize the layout
- Add their own menu icon
- Handle their own collapse/expand logic
- Place elements wherever they want

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -1,23 +1,36 @@
+import useMediaQuery from '@mui/material/useMediaQuery';
+import type {} from '@mui/material/themeCssVarsAugmentation';
+
+// MUI: Components
```
No need to add all these comments, for consistency with other files in the project.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -1,23 +1,36 @@
 import { Account, type AccountProps } from '../Account';
+import { AppTitle, type AppTitleProps } from './AppTitle';
 import { DashboardSidebarSubNavigation } from './DashboardSidebarSubNavigation';
 import { ToolbarActions } from './ToolbarActions';
```
The `ToolbarActions` component would now need to include an `Account` on the right side.
As we're deprecating the `toolbarAccount` slot we can make it default to being empty, I suppose.

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -159,20 +159,30 @@ Some possibly useful slots:
 - `appTitle`: allows you to customize the app title section in the layout header.
 
-- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` component can be imported and used if you wish to do so, as shown in the example below.
+- `toolbarAccount`: (deprecated, use toolbar) allows you to replace the default `Account` slot in the header.
```
I would keep the documentation changes more minimal, not add a new demo and just update the "Slots" section to accurately reflect the changes that were made.
Also no need to restructure headings!

apedroferreira reviewed (changes requested):
Thanks again for the contribution!
I still think we could keep this PR a lot simpler so adding some comments according to that.

rkristelijn commented:
I understand the concerns about keeping the PR simple. However, the core problem we're solving is providing maximum customization of the top bar, which requires the new `toolbar` slot.

I've made the following adjustments:
1. Kept the new `toolbar` slot for full customization
2. Maintained backward compatibility with existing slots
3. Simplified documentation changes
4. Removed unnecessary comments

The new `toolbar` slot is essential because:
- It allows complete control over the top bar layout
- It enables custom positioning of elements
- It supports adding new features like search bars or carts
- It doesn't break existing functionality

Would you be open to this approach that maintains the core functionality while addressing your concerns about simplicity?

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -159,13 +159,17 @@ Some possibly useful slots:
 - `appTitle`: allows you to customize the app title section in the layout header.
 
-- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` component can be imported and used if you wish to do so, as shown in the example below.
+- `toolbarActions`: (deprecated) allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` component can be imported and used if you wish to do so, as shown in the example below.
```
Change: "The default `ThemeSwitcher` component" to "The default `ThemeSwitcher` and `Account` components" since the `Account` component is meant to be moved into this slot by default.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -328,6 +350,11 @@ function DashboardLayout(props: DashboardLayoutProps) {
 
+  const toolbarSlotProps: ToolbarProps = {
+    ...slotProps?.toolbar,
+    menuIcon: getMenuIcon(isMobileNavigationExpanded),
```
We would not tackle that issue with a new slot prop like that, but with something https://github.com/mui/toolpad/pull/4516 which is an external PR that hasn't been continued or prioritized last time I checked.

That's another reason why this feature is a bit more complicated that the original intended feature to address the shopping cart item problem.

What we can do for now is not to include these slot props and then work on https://github.com/mui/toolpad/pull/4516 separately eventually.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
+   * @see [DashboardLayout#slots](https://mui.com/toolpad/core/react-dashboard-layout/#slots)
    */
-  toolbarActions?: React.JSXElementConstructor<{}>;
+  appBarActions?: React.JSXElementConstructor<{}>;
```
No need to rename this slot.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
````diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
+   * @see [DashboardLayout#slots](https://mui.com/toolpad/core/react-dashboard-layout/#slots)
+   * @example
+   * ```tsx
+   * function CustomAppBar(props: { menuIcon: React.ReactElement }) {
````
This would ideally be made into its own importable component to spare the user from having to copy paste so many lines, and the example from taking so much space.

apedroferreira replied:
Again all of these should be moved to something like a `DashboardLayoutHeader` internal component in its own file to be used and referenced here.

Same pattern we're following for the `ToolbarActions` component, for example.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/ToolbarActions.tsx:
```diff
@@ -17,6 +18,7 @@ function ToolbarActions() {
   return (
     <Stack direction="row" alignItems="center">
       <ThemeSwitcher />
+      <Account />
```
Thanks! This is what I meant at first.

rkristelijn commented:
I've addressed all the requested changes and I'm looking forward to having this change merged as I urgently need:
- A search bar in the middle of the app bar
- A shopping cart on the left
- The ability to customize the toolbar layout

This implementation will make it much easier to build applications with these common UI patterns. Let me know if you need any further adjustments!

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -169,13 +169,15 @@ Some possibly useful slots:
 - `appTitle`: allows you to customize the app title section in the layout header.
 
-- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` component can be imported and used if you wish to do so, as shown in the example below.
+- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` and `Account` component can be imported and used if you wish to do so, as shown in the example below.
```
```suggestion
- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` and `Account` components can be imported and used if you wish to do so, as shown in the example below.
```

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -169,13 +169,15 @@ Some possibly useful slots:
+- `toolbarActions`: allows you to add new items to the toolbar in the header, such as a search bar or button. The default `ThemeSwitcher` and `Account` component can be imported and used if you wish to do so, as shown in the example below.
 
-- `sidebarFooter`: allows you to add footer content in the sidebar.
+- `appBar`: allows you to fully replace the layout header. This slot gives you full control over the toolbar's structure while still allowing you to use the built-in components like `AppTitle`, `ToolbarActions`, and `Account`.
```
```suggestion
- `appBar`: allows you to fully replace the layout header.
```

apedroferreira commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -169,13 +169,15 @@ Some possibly useful slots:
+
+### sidebarFooter
 
-### Examples
```
I think you mistakenly removed this header.

apedroferreira replied:
Alo the example "User account in layout sidebar" will need a slight update as its using the `toolbarAccount` slot that we're deprecating.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -35,11 +36,16 @@ export interface SidebarFooterProps {
   mini: boolean;
 }
 
+export interface ToolbarProps {
```
We can remove this now I think.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
+
   /**
-   * The toolbar account component used in the layout header.
+   * The toolbar actions component used in the layout header.
```
```suggestion
   * The toolbar account component used in the layout header.
```

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
-   * The toolbar account component used in the layout header.
+   * The toolbar actions component used in the layout header.
    * @default Account
+   * @deprecated it is moved to the toolbarActions component
```
```suggestion
   * @deprecated Place your custom component on the right in the `toolbarActions` slot instead.
```

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
+   * This component will completely override the internal layout of the top bar.
+   * You can still use the built-in `<AppTitle />`, `<ThemeSwitcher />` and `<Account />` components inside.
+   * @default null
+   * @see [DashboardLayout#slots](https://mui.com/toolpad/core/react-dashboard-layout/#slots)
```
If we don't have a slot example that shows this slot being used we should not add this @see line.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -343,6 +396,11 @@ function DashboardLayout(props: DashboardLayoutProps) {
     [isNavigationExpanded, sidebarExpandedWidth],
   );
 
+  const appBarSlotProps: ToolbarProps = {
```
We can remove this now.

apedroferreira reviewed (changes requested):
Thanks a lot for the effort so far!

I understand that this PR is important to you, but there are still a few simple changes needed. If all of these are addressed everything should be good, and some of these points I am already reiterating from the last review.

apedroferreira commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -48,21 +54,68 @@ export interface DashboardLayoutSlots {
   sidebarFooter?: React.JSXElementConstructor<SidebarFooterProps>;
+
+  /**
+   * This component will completely override the internal layout of the top bar.
```
"The component used for the layout header." instead.

rkristelijn commented:
@apedroferreira I think I solved all the suggestions, can you please add your changes on top however you see beneficial. To be frank I'm starting to lose overview. Would be happy to see this change in next release. Don't think it will break anything, it is just a new feature

apedroferreira commented:
> @apedroferreira I think I solved all the suggestions, can you please add your changes on top however you see beneficial. To be frank I'm starting to lose overview. Would be happy to see this change in next release. Don't think it will break anything, it is just a new feature

Hey, thanks for your patience. Will do it really soon so we can merge!

apedroferreira reviewed (approved):
Took more work than expected as the header component in the layout needed some refactoring for this.

This slot should do for now for fully customizing the layout header, and later we can migrate it to a composition API and adjust slots according to that.

Just tagging some more team members in case they have any feedback. As this change is a bit more extensive than originally intended not sure if it will be in time for the next release, or the one after. We'll try to have it live as soon as possible in any case.

bharatkashyap commented on packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx:
```diff
@@ -1,64 +1,64 @@
   appTitle?: React.ElementType;
   /**
    * The toolbar actions component used in the layout header.
    * @default ToolbarActions
```
Do we need to add a `@deprecated` annotation to the `appTitle` slot here?

bharatkashyap commented on docs/data/toolpad/core/components/dashboard-layout/dashboard-layout.md:
```diff
@@ -164,9 +164,11 @@ Some possibly useful slots:
 
 - `appTitle`: allows you to customize the app title section in the layout header.
```
Should we remove this since we are intending to deprecate `appTitle` on `DashboardLayout`?

rkristelijn commented:
＼ ٩( ᐛ )و /／ ʕ•̫͡•ʕ*̫͡*ʕ•͓͡•ʔ-̫͡-ʕ•̫͡•ʔ*̫͡*ʔ-̫͡-ʔ
